src/extra/recipes/vs.py

```diff
@@ -0,0 +1,890 @@
+from typing import Optional, Union, Dict, List, Tuple, Any
+from dataclasses import dataclass, asdict, is_dataclass
+
+import itertools
+from functools import partial
+from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor
+import logging
+
+import numpy as np
+from extra_data import open_run, by_id, DataCollection, KeyData
+import h5py
+import xarray as xr
+import pandas as pd
+
+from .base import SerializableMixin
+
+from extra.components import AdqRawChannel, XrayPulses, XGM, Scan
+
+from scipy.signal import csd, fftconvolve, firwin, kaiserord, argrelmin
+
+def search_offset(trace: np.ndarray, sigma: float=20) -> int:
+    """
+    Find highest peaks in the 1D trace.
+
+    Args:
+      trace: Trace.
+      sigma: Sigma of the Gaussian to convolve.
+
+    Returns: Offset with some slack.
+    """
+    # apply it to the data
+    from scipy.ndimage import gaussian_filter1d
+    smoothened = gaussian_filter1d(trace, sigma=sigma, mode="nearest")
+    peak_idx = np.argmax(smoothened)
+    return peak_idx - 200
+
+def fwhm(x: np.ndarray, y: np.ndarray) -> float:
+    """Return the full width at half maximum of x."""
+    # half maximum
+    half_max = 0.5*np.amax(y)
+    idx = np.where(y > half_max)[0]
+    if len(idx) < 2:
+        return 999
```

Comment:
Would maybe 0 or -1 be a more obvious signal? Or even throwing an exception? 999 is a plausible (if unlikely) FWHM.

Reply:
Agreed. I don't want to throw an exception, as this is for validation only, but using -1 is a good idea.